[WIP] Switch explorer rendering to use react-virtualised

Akin909 opened the pull request:
An issue I noticed with the explorer was that on reaching the lowest boundaries of the visible files/folders the scroll position would flicker on each move downwards making it hard/impossible to move to files which are out of view.

Having attempted to use/tweak the `scrollIfNeeded` function with no luck I've refactored the explorer to use react virtualize instead which I believe was mentioned at some point was an eventual change we wanted to make.

Todo:
- [ ] fix scroll issues when folder(s) is/are expanded - at present the main hold up with this PR is that if a folder is expanded react virtualised seems maybe to not know how to scroll since ?the dimensions are off?

bryphe commented on browser/src/Services/Explorer/ExplorerView.tsx:
```diff
@@ -269,11 +265,72 @@ export interface IExplorerViewProps extends IExplorerViewContainerProps {
+
+        switch (true) {
+            case renameInProgress:
+                return 35
```
Hmm, how do we determine these sizes? I'm wondering if they are dependent on the `ui.fontSize` setting? I'm just worried about a bug where different font sizes cause rendering quirks.

Akin909 replied:
I actually fiddled with the sizes here and tbh the css with `virtualized` involved is more or less opaque to me but I didn't notice a big difference using the values which were trial and error as a result of that process, I've tweaked it however to use `ui.fontSize` to derive the size in case

bryphe commented on browser/src/Services/Explorer/ExplorerView.tsx:
```diff
@@ -289,36 +346,45 @@ export class ExplorerView extends React.PureComponent<IExplorerViewProps, {}> {
-                            </div>
+                            <ExplorerContainer className="explorer">
+                                <AutoSizer>
+                                    {({ height, width }) => (
```
Nice! Thanks for wiring this all up 👍

bryphe commented:
> An issue I noticed with the explorer was that on reaching the lowest boundaries of the visible files/folders the scroll position would flicker on each move downwards making it hard/impossible to move to files which are out of view.

Ah ya, I hit this too - didn't get a chance to log an issue, but it definitely was a problem! Thanks so much for fixing it!

> Having attempted to use/tweak the scrollIfNeeded function with no luck I've refactored the explorer to use react-virtualized instead which I believe was mentioned at some point was an eventual change we wanted to make, which has the added bonus of providing methods that can be used to control scrolling to selected elements.

Awesome!! 💯 Thanks for going to those lengths to fix it. This will be much better long-term too, especially for large directories with tons of files - we only have to render whats on the screen.

> This change also sets up the searchInExplorer functionality I'm planning by allowing for programmatically focusing search of a particular element.

Can't wait! 😄 

Overall the change looks great! I had one small concern around how we calculate the row heights - I'm not sure how it works with different font sizes, so I wanted to see what you thought there.

Thanks @Akin909 !

Akin909 commented:
@bryphe made the changes you suggested re row height 👍 hopefully this is good to go

bryphe commented on browser/src/Services/Explorer/ExplorerView.tsx:
```diff
@@ -269,11 +266,77 @@ export interface IExplorerViewProps extends IExplorerViewContainerProps {
+        const isSelected = id === selectedId
+
+        const PADDING = 10
+        const fontSize = parseInt(this.props.fontSize, 10)
```
Hmm, I wonder if this handles the various cases we allow for fontsize? For example, it looks like `11px`, `11pt`, `11em` would all get handled the same way here, but they would actually have very different _pixel_ sizes. I think what we'd want instead is the measured font size, or some more deterministic way to get the size of the cells.

Akin909 replied:
@bryphe I used the menu component which has a static value of `20` as my initial reference point for implementing this and I haven't noted any issue with the value we pass in there, I also tried various sizes with a static value for `rowHeight` locally and didn't get any issues re. different units I can look at converting em to px as for pt I really hope ppl aren't giving units in pt as that'll be very problematic on multiple fronts.

bryphe commented on browser/src/Services/Explorer/ExplorerView.tsx:
```diff
@@ -269,11 +266,77 @@ export interface IExplorerViewProps extends IExplorerViewContainerProps {
+
+        switch (true) {
+            case renameInProgress:
+                return rowSize * 1.5
```
I'm a bit concerned about these magic numbers here - `1.5` and `2.8`. If someone changes the styling of the node or the creation, they might not realize they need to update these here, and get difficult to find bugs. Along the lines of the above comment - is there a more deterministic way we can get these sizes?

Akin909 commented:
Switched to using the virtualized `CellMeasurer` which handles row sizing dynamically (not the most performant solution but the *only* one I could find after a great many different attempts)

bryphe commented:
Wow, thanks for figuring out how to get accurate row sizes, @Akin909 ! Awesome.

bryphe commented:
Pulled the latest changes to check it out on Windows - I used the `Paint Performance` tool. I wanted to make sure we weren't aggressively re-painting the explorer while typing, which we aren't 👍 

I think the reason this test is failing is because the paint happens just a bit later, so it gets picked up the test. The best bet is to simply turn off the sidebar / explorer in that test's config:
https://github.com/onivim/oni/blob/cd87a1f201ea8b20659c7d76d3a57464ce186cb8/test/ci/PaintPerformanceTest.config.js#L5

(That will reduce the noise - the test is really focused on the _editor surface_). Note that we disable the statusbar, tabs, and animations for the same reason 😄

Akin909 commented:
@bryphe i've updated the test for this and its good to go

CrossR commented:
Code looks good, and the stuttering is gone, but when scrolling fast I get this slight lag between the selected item and the highlight. 

Bad gif example in the Oni repo just holding down `j`:

![issue](https://user-images.githubusercontent.com/10038688/41369887-163cb0a6-6f3e-11e8-8212-aedb30f37bc2.gif)

That said its still a lot better than the current flickering, do you want to address the lag issue in a separate PR?

Akin909 commented:
Gonna stick a work in progress on this as I've just done some more local testing and theres a weird sizing issue when trying to create a folder in the root dir :S didn't realise 🤦‍♂️

CrossR commented:
I was using the latest build at the time, so afc992fc4fa27a06cfd8207ac54fe43dc4dd6827.

I'll try again using my own build rather than grabbing the AppVeyor one though and see if its different.

bryphe commented:
Thanks @Akin909 and @CrossR for the all the testing here! 💯

Akin909 commented:
@CrossR could you give this PR a try especially re creating and renaming and moving files, I've kind of lost perspective on it thing its a big improvement (re. the stutter) and havent noted any bugs today but previously noted a resizing issue which seems to have resolved itself.....

keforbes commented:
Hey @Akin909, it looks like this PR has a branch conflict now.  Can you take a look?

CrossR commented:
I'm still getting some weirdness with the scroll sticking to the bottom.

If I scroll past the bottom of the menu, the bar sticks to the bottom for the entire time now. There is also some jumping still where the bar vanishes for a bit and then reappears after letting go of the scroll button.

Another thing I've noticed is that on inserting items sometimes the spacing is off.
Repro: 
 - Open Oni with Oni repo. 
 - `Ctrl-h` to highlight explorer, 
 - `G` to move to bottom, 
 - `Ctrl-e` to make a new file, and set name to `a.txt`.
 - New file is added fine, but weird spacing occurs:

![image](https://user-images.githubusercontent.com/10038688/44875521-58d18c80-ac96-11e8-8e94-c6fa243cfe7b.png)

Its up to you if you want to sort them in this PR or in a follow up, neither are as bad as the flickering, so I'd understand getting this in and sorting the bugs later(in a smaller more contained PR).

Let me know and if you decide on that, I can approve the PR. No point doing that now since there is a merge conflict.

Akin909 commented:
@CrossR thanks for having another look was reeeeeally stuck with the bug you noticed with the spacing I've switched the strategy to a more brute force one instead of only trying to update the affected nodes when something like a rename etc happens I instead now trigger a redraw/update of all the visible nodes so should avoid random ones here or there not being measured properly.

Re holding the button and the bar disappearing when is that happening? I haven't noted it since I switched to having the bar stick on the bottom although, one of the limitations of using virtualized and the cellMeasurer component from the docs is

> Since this component measures one cell at a time to determine it's width/height, it will likely be slow if a user skips many rows (or columns) at once by scrolling with a scrollbar or via a scroll-to-cell prop. There is (unfortunately) no workaround for this performance limitation at the moment.

So without finding a faster library or a different implementation button mashing could always lead to this although I there might be little ways to optimise this or hide it at least. Although the library authors have been talking about a new version for a while now so we'll see.

Anyway if you get a moment can you try it again to see if the weird spacing bug is still there(locally I havent noted it anymore), I never found a consistent way to reproduce it (your repro doesnt always bring it on nor have I found it to consistently reproduce)

CrossR commented:
>Re holding the button and the bar disappearing when is that happening? I haven't noted it since I switched to having the bar stick on the bottom although, one of the limitations of using virtualized and the cellMeasurer component from the docs is

Seems to still happen when I scroll rapidly in the Oni repo:

![gif](https://thumbs.gfycat.com/ImperturbableDiscreteBellsnake-size_restricted.gif)

I can't seem to repro the other bug now though, and I tried 10-15 times whereas before it was happening maybe every 2/3 times I tried.

Akin909 commented:
@CrossR thanks for the GIF that helps a lot 💯 , re that issue it relates to the quote above re performance and the strategy were using basically the library cant keep up with the holding the button

As far as the docs say there is no clear way round it I'm hoping future releases improve performance and we get a fix for free if not down the line we might have to switch to a different strategy I have no idea what, the main value of the library is for scrolling v. long lists not quite what were doing which is scrolling to indexes very fast using keyboard input
